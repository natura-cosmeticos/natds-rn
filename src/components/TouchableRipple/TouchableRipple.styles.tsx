import styled from 'styled-components/native';
import {
  getColorPrimary,
  getColorSecondary,
  getColorHighlight,
  Theme,
} from '../../common/themeSelectors';
import { TouchableRippleProps, TouchableRippleColors } from './TouchableRipple';

const getBackgroundColor = (color: TouchableRippleColors, theme: Theme) => {
  let colorCode;

  switch (color) {
    case 'secondary':
      colorCode = getColorSecondary(theme);
      break;
    case 'primary':
      colorCode = getColorPrimary(theme);
      break;
    case 'highlight':
      colorCode = getColorHighlight(theme);
      break;
    default:
      colorCode = getColorPrimary(theme);
  }

  return colorCode;
};

type RippleProps = Pick<TouchableRippleProps, 'theme' | 'size' | 'color'>

export const RippleWrapper = styled.View<Pick<TouchableRippleProps, 'hideOverflow'>>(({ hideOverflow }) => ({
  bottom: 0,
  elevation: 999,
  left: 0,
  overflow: hideOverflow ? 'hidden' : 'visible',
  position: 'absolute',
  right: 0,
  top: 0,
  zIndex: 999,
}));

export const Ripple = styled.View<Required<RippleProps>>(({
  theme, size, color,
}) => ({
  backgroundColor: getBackgroundColor(color, theme),
  borderRadius: size / 2,
  height: size,
  position: 'absolute',
  width: size,
}));
