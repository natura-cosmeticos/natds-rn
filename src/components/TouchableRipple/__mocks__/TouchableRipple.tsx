import React from 'react';
import { TouchableWithoutFeedback } from 'react-native';

/* eslint-disable */
export const TouchableRipple = ({ children, ...props }) => {
  return (
    <TouchableWithoutFeedback {...props}>
      {typeof children === 'function' ? '[Child as a function]' : children}
    </TouchableWithoutFeedback>
  );
};
