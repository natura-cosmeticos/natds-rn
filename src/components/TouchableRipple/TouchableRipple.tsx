/* eslint-disable max-statements, max-lines, no-underscore-dangle */

import React, { ReactElement, useRef, useState } from 'react';
import {
  TouchableWithoutFeedback,
  Animated,
  Easing,
  Platform,
  findNodeHandle,
  UIManager,
} from 'react-native';
import { Container, Ripple } from './TouchableRipple.styles';
import { Theme } from '../../common/themeSelectors';

export type TouchableRippleColors = 'primary' | 'secondary' | 'highlight';

export interface TouchableRippleProps {
  /**
   * Children component
   */
  children?: ReactElement;
  /**
   * Ripple color: `primary` | `secondary` | `highlight`
   * @default: `primary`
   */
  color?: TouchableRippleColors;
  /*
   * Deactivates the palpable effect, will not call the callback function when pressing;
   */
  disabled?: boolean;
  /**
   * Controls if the ripple should overflow the content of not
   */
  hideOverflow?: boolean;
  /**
   * Size of the children, the ripple will have the double of this size.
   */
  size: number;
  /**
   * Will be called as soon the ripple animation start
   */
  onPress?: () => void;
  /**
   * Optional testID
   */
  testID?: string;
  /**
   * Optional testID
   */
  theme?: Theme;
}

export const getChildrenPosition = (ref, setPosition: Function) => {
  if (ref && ref._children && ref._children[0]) {
    UIManager.measure(
      findNodeHandle(ref._children[0]) || 0,
      (elX, elY, width, height) => {
        setPosition({ left: elX + (width / 2), top: elY + (height / 2) });
      },
    );
  }
};

interface State {
  left: string
  top: string
}

export const TouchableRipple = ({
  children,
  color = 'primary',
  disabled = false,
  size,
  onPress,
  testID = 'touchable-ripple',
  hideOverflow = false,
}: TouchableRippleProps) => {
  const maxOpacity = 0.16;
  const animationDuration = 255;
  const isNative = Platform.OS !== 'web';

  const scaleValue = useRef(new Animated.Value(0)).current;
  const opacityValue = useRef(new Animated.Value(0)).current;
  const rippleSize = size * 2;

  const [position, setPosition] = useState<State>({
    left: '50%',
    top: '50%',
  });

  function resetAnimations() {
    scaleValue.setValue(0);
    opacityValue.setValue(0);
  }

  function onPressOutTouchable() {
    if (onPress) {
      Animated.timing(opacityValue, {
        duration: animationDuration,
        toValue: 0,
        useNativeDriver: isNative,
      }).start(resetAnimations);
    }
  }

  function onPressTouchable() {
    if (onPress) {
      resetAnimations();
      Animated.parallel([
        Animated.timing(scaleValue, {
          duration: animationDuration,
          easing: Easing.bezier(0.0, 0.0, 0.2, 1),
          toValue: 1,
          useNativeDriver: isNative,
        }),
        Animated.timing(opacityValue, {
          duration: animationDuration,
          toValue: maxOpacity,
          useNativeDriver: isNative,
        }),
      ]).start();
    }
  }

  return (
    <TouchableWithoutFeedback
      disabled={disabled}
      onPressIn={onPressTouchable}
      onPressOut={onPressOutTouchable}
      onPress={onPress}
      delayPressOut={animationDuration}
      testID={testID}
    >
      <Container
        hideOverflow={hideOverflow}
        ref={ref => getChildrenPosition(ref, setPosition)}
      >
        {children}
        <Ripple
          as={Animated.View}
          size={rippleSize}
          color={color}
          style={{
            left: position.left,
            opacity: opacityValue,
            top: position.top,
            transform: [
              { translateX: -(rippleSize / 2) },
              { translateY: -(rippleSize / 2) },
              { scale: scaleValue },
            ],
          }}
        />
      </Container>
    </TouchableWithoutFeedback>
  );
};
