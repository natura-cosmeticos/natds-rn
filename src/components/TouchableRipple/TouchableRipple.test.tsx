import React, { createRef } from 'react';
import { View, Text, Animated, UIManager } from 'react-native';
import { fireEvent } from '@testing-library/react-native';
import { TouchableRipple, TouchableRippleProps, getChildrenPosition } from './TouchableRipple';
import { render as customRender } from '../../../test/testHelpers';

jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');

  RN.UIManager.measure = (node, cb) => cb(1, 2, 10, 10);

  Object.defineProperty(RN, 'findNodeHandle', {
    get: jest.fn(() => () => 1),
    set: jest.fn(),
  });

  RN.Animated.parallel = jest.fn();
  RN.Animated.timing = jest.fn();

  return RN;
});

jest.mock(
  'react-native/Libraries/Components/Touchable/TouchableWithoutFeedback',
  () => 'TouchableWithoutFeedback',
);

const defaultProps = {
  children: <View />,
  onPress: () => null,
  size: 24,
};

const renderTouchableRipple = (
  props?: Partial<TouchableRippleProps>,
) => customRender<TouchableRippleProps>(
  <TouchableRipple {...defaultProps} {...props} >
    <Text>I have a ripple!</Text>
  </TouchableRipple>,
);

describe('TouchableRipple component', () => {
  it('Should render touchable ripple correctly', () => {
    const { toJSON } = renderTouchableRipple();

    expect(toJSON()).toMatchSnapshot();
  });

  it('Should render touchable ripple with color', () => {
    const { toJSON } = renderTouchableRipple({ color: 'secondary' });

    expect(toJSON()).toMatchSnapshot();
  });

  it('Should render touchable ripple with disabled', () => {
    const { toJSON } = renderTouchableRipple({ disabled: true });

    expect(toJSON()).toMatchSnapshot();
  });

  it('Should render touchable ripple with hideOverflow', () => {
    const { toJSON } = renderTouchableRipple({ hideOverflow: true });

    expect(toJSON()).toMatchSnapshot();
  });

  it('it should call onPress on touchable press', () => {
    const onPress = jest.fn();
    const { queryByTestId } = renderTouchableRipple({ onPress });

    const touchable = queryByTestId('touchable-ripple');

    if (touchable) {
      fireEvent.press(touchable);
    }

    expect(onPress).toHaveBeenCalled();
  });

  it('it should run the grow animation on press in', () => {
    const timing = jest.fn();
    const start = jest.fn();

    Animated
      .parallel
      // @ts-ignore
      .mockImplementation(jest.fn())
      .mockReturnValue({ start });

    Animated
      .timing
      // @ts-ignore
      .mockImplementation(timing);

    const { queryByTestId } = renderTouchableRipple();
    const touchable = queryByTestId('touchable-ripple');

    if (touchable) {
      fireEvent(touchable, 'onPressIn');
    }

    expect(start).toHaveBeenCalled();
    expect(timing).toHaveBeenCalledTimes(2);
  });

  it('it should run the shrink animation on press out', () => {
    const timing = jest.fn();
    const start = jest.fn();

    Animated
      .timing
      // @ts-ignore
      .mockImplementation(timing)
      // @ts-ignore
      .mockReturnValue({ start });

    const { queryByTestId } = renderTouchableRipple();

    const touchable = queryByTestId('touchable-ripple');

    if (touchable) {
      fireEvent(touchable, 'onPressOut');
    }

    expect(start).toHaveBeenCalled();
  });
});

describe('getChildrenPosition', () => {
  it('should set ripple position based on ref', () => {
    const ref = {
      _children: [{}],
    };

    const setPositionMock = jest.fn();

    getChildrenPosition(ref, setPositionMock);

    expect(setPositionMock).toHaveBeenCalledWith({
      left: 6,
      top: 7,
    });
  });
});
