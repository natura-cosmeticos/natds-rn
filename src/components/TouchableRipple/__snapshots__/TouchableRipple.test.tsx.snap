// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TouchableRipple component Should render touchable ripple correctly 1`] = `
<TouchableWithoutFeedback
  delayPressOut={100}
  disabled={false}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  testID="touchable-ripple"
>
  <View
    hideOverflow={false}
    style={
      Array [
        Object {
          "overflow": "visible",
        },
      ]
    }
  >
    <Text>
      I have a ripple!
    </Text>
    <View
      color="primary"
      size={48}
      style={
        Object {
          "backgroundColor": "#f4ab34",
          "borderRadius": 24,
          "height": 48,
          "left": "50%",
          "opacity": 0,
          "position": "absolute",
          "top": "50%",
          "transform": Array [
            Object {
              "translateX": -24,
            },
            Object {
              "translateY": -24,
            },
            Object {
              "scale": 0,
            },
          ],
          "width": 48,
          "zIndex": 999,
        }
      }
    />
  </View>
</TouchableWithoutFeedback>
`;

exports[`TouchableRipple component Should render touchable ripple with color 1`] = `
<TouchableWithoutFeedback
  delayPressOut={100}
  disabled={false}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  testID="touchable-ripple"
>
  <View
    hideOverflow={false}
    style={
      Array [
        Object {
          "overflow": "visible",
        },
      ]
    }
  >
    <Text>
      I have a ripple!
    </Text>
    <View
      color="secondary"
      size={48}
      style={
        Object {
          "backgroundColor": "#ff6b0b",
          "borderRadius": 24,
          "height": 48,
          "left": "50%",
          "opacity": 0,
          "position": "absolute",
          "top": "50%",
          "transform": Array [
            Object {
              "translateX": -24,
            },
            Object {
              "translateY": -24,
            },
            Object {
              "scale": 0,
            },
          ],
          "width": 48,
          "zIndex": 999,
        }
      }
    />
  </View>
</TouchableWithoutFeedback>
`;

exports[`TouchableRipple component Should render touchable ripple with disabled 1`] = `
<TouchableWithoutFeedback
  delayPressOut={100}
  disabled={true}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  testID="touchable-ripple"
>
  <View
    hideOverflow={false}
    style={
      Array [
        Object {
          "overflow": "visible",
        },
      ]
    }
  >
    <Text>
      I have a ripple!
    </Text>
    <View
      color="primary"
      size={48}
      style={
        Object {
          "backgroundColor": "#f4ab34",
          "borderRadius": 24,
          "height": 48,
          "left": "50%",
          "opacity": 0,
          "position": "absolute",
          "top": "50%",
          "transform": Array [
            Object {
              "translateX": -24,
            },
            Object {
              "translateY": -24,
            },
            Object {
              "scale": 0,
            },
          ],
          "width": 48,
          "zIndex": 999,
        }
      }
    />
  </View>
</TouchableWithoutFeedback>
`;

exports[`TouchableRipple component Should render touchable ripple with hideOverflow 1`] = `
<TouchableWithoutFeedback
  delayPressOut={100}
  disabled={false}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  testID="touchable-ripple"
>
  <View
    hideOverflow={true}
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
      ]
    }
  >
    <Text>
      I have a ripple!
    </Text>
    <View
      color="primary"
      size={48}
      style={
        Object {
          "backgroundColor": "#f4ab34",
          "borderRadius": 24,
          "height": 48,
          "left": "50%",
          "opacity": 0,
          "position": "absolute",
          "top": "50%",
          "transform": Array [
            Object {
              "translateX": -24,
            },
            Object {
              "translateY": -24,
            },
            Object {
              "scale": 0,
            },
          ],
          "width": 48,
          "zIndex": 999,
        }
      }
    />
  </View>
</TouchableWithoutFeedback>
`;
